### ISSUE TITLE
Missing check for spike times before t_start in BlackRockIO

### ISSUE THREAD

**Issue by bjoern1001001** [NeuralEnsemble/python-neo#420]
Trying to load real datasets with the latest BlackRockIO, I noticed that some of them would raise the error that spike times are before t_start or after t_stop respectively. This seems to happen because these t_start and t_stop are determined based on the nsX files available. However, in rare cases also spike times just outside of the nsX recording time range can be present, therefore leading to a 'ValueError: The first spike (..) is before t_start (..)'. When no nsX file is available the IO works just fine.

So @JuliaSprenger 's and my suggested solution is to explicitly check for this and correct t_start and t_stop for the SpikeTrains accordingly by setting it to the min_nev_time (first spike observed) or max_nev_time (last spike observed) respectively.

I'd try implementing this myself tomorrow if there are no other suggestions.

**Comment by bjoern1001001** [comment]
So I tried implementing this, but found a few difficulties, especially regarding the time_slices parameter in read_block. I'd like to ask you, @samuelgarcia, what you think should be done here.

What happens in BaseFromRawIO is that all t_starts (for AnalogSignals and SpikeTrains) are the same. However, because spike times can be before this t_start, that yields a few problems:
It's not possible to set all t_start to the smallest existing value (in this case the first spike) because this would shift the AnalogSignals towards earlier times which would be incorrect.
On the other hand setting SpikeTrain t_start to a lower value independently makes it impossible to find those spikes when setting the time_slices parameter in read_block, because the check whether the values are correct is done earlier.
Apart from that it's not always known when segments start or stop to people working with the data. In order to prevent frequent errors from happening, earlier versions of BlackRockIO would automatically set n_starts and n_stops to correct values, i.e. too small values to the minimal available time and too high values to the highest possible value. This also holds true, if 'None' was specified for either of them. So @JuliaSprenger and I think this behavior should be kept.

To sum all of this up, times should probably be corrected automatically and this should happen independently for AnalogSignals and SpikeTrains, so AnalogSignals can keep their correct t_start and the spikes occurring earlier than this can be read and shown as well.

**Comment by samuelgarcia** [comment]
Hi Bjorn.
time_slices in read_block() is really ambiguous, 
I would prefer to discuss this case in #374 to avoid mixing all discusions.

In the new master, in BaseFromRawIO, I have ported more or less the behiavior of old BlackrockIO.
I don't really like this because there are special where it do not make sens for instance when all segment start at 0s.
Since there are not garanty on some IO of global time time slice at block level should be avoid in the raw API
I think you should have something more clear.

My proposal is to remove:
  * to implemented a new  lazy load system with proxy object see  #406
  * make lazy load by default
  * implement time slice on proxy object only.
  * remove totally time_slice for read_segment() and time_slices for read_block().
So the new way for time slice could be:
```python
reader = BlackrockIO(filename='XXX', lazy=True)
block = reader.read_block()
proxysig = block.sements[15].analogsignals[18]
anasig = proxysig.load(time_slice=[1,5])
```
I like it more, it is not ambiguous and more consistent.
We need strong discussion about this before on #406 
In any case I won't have time to code theses proxy object before january.


For the problem of spike that happen before signal, this can be fix in BlackrockRawIO easily.
RawIO API propose both _segment_t_start() and _get_signal_t_start().
In BlackrockRawIO by mistake there are the same values. We should put the min(first_spike, _get_signal_t_start) to _segment_t_start and this whould solve the problem without changing line in FromBaseIO.

Do you prefer I do it or do you want to send a PR ?

Best


Samuel

**Comment by samuelgarcia** [comment]
Fix by #421.
